File: src/main/java/com/github/frankfuenmayor/flutterhelper/settings/SettingsPanel.java
package com.github.frankfuenmayor.flutterhelper.settings;

import com.github.frankfuenmayor.flutterhelper.buildrunner.BuildRunnerAnnotation;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.List;

import com.github.frankfuenmayor.flutterhelper.buildrunner.Icons;
import com.intellij.openapi.options.Configurable;
import com.intellij.util.ui.AbstractTableCellEditor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SettingsPanel implements Configurable {
    JPanel rootPanel;
    private JTable annotationsTable;
    private JButton button1;

    AnnotationsTableModel model;

    public SettingsPanel() {
        List<@NotNull BuildRunnerAnnotation> initialAnnotations = SettingsService.Companion.getInstance().getState().annotations;

button1.setIcon(Icons.Add);

        model = new AnnotationsTableModel(initialAnnotations);
        annotationsTable.setModel(model);
        annotationsTable.setCellEditor(new AbstractTableCellEditor() {
            @Override
            public Object getCellEditorValue() {
                return null;
            }

            @Override
            public Component getTableCellEditorComponent(JTable table, Object value, boolean isSelected, int row, int column) {
                return new JTextField();
            }
        });

        button1.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                model.addRow();
            }
        });
    }

    @Override
    public @Nullable JComponent createComponent() {
        return rootPanel;
    }

    public boolean isModified() {
        return SettingsService.Companion.getInstance().getState().annotations.size() != model.getAnnotations().size();
    }

    @Override
    public void apply() {
        SettingsService
                .getInstance()
                .setAnnotations(model.getAnnotations());
    }

    @Override
    public void reset() {
        model.reset();
    }

    @Override
    public String getDisplayName() {
        return "Flutter Helper";
    }
}

File: src/main/kotlin/com/github/frankfuenmayor/flutterhelper/settings/SettingsPanel.kt
package com.github.frankfuenmayor.flutterhelper.settings

import com.intellij.openapi.options.Configurable
import com.intellij.openapi.project.ProjectManager
import java.awt.BorderLayout
import javax.swing.*


class SettingsPanel : Configurable {
    private val settings =
        SettingsService.getInstance().state

    override fun createComponent(): JComponent = JPanel().apply {

        layout = BorderLayout()

        add(JLabel("Dart build_runner annotations"), BorderLayout.NORTH)
        val model = AnnotationsTableModel(settings);
        val table = JTable(model)
        add(JScrollPane(table), BorderLayout.CENTER)
    }

    override fun isModified(): Boolean {
        return false
    }

    override fun apply() {
//        val settings = PluginSettingsService.getInstance(com.intellij.openapi.project.ProjectManager.getInstance().defaultProject).state
//        settings.option1 = option1Field?.text ?: ""
//        settings.option2 = option2Checkbox?.isSelected ?: false
    }

    override fun reset() {
//        val settings = PluginSettingsService.getInstance(com.intellij.openapi.project.ProjectManager.getInstance().defaultProject).state
//        option1Field?.text = settings.option1
//        option2Checkbox?.isSelected = settings.option2
    }

    override fun getDisplayName(): String = "Flutter Helper"
}